File: fix_regions.py
"""Assign Bundesland to _all_points entries using a bounding-box lookup (no external shapefiles needed)."""
import json, sys

# Rough bounding boxes for German Bundesländer (lat_min, lat_max, lon_min, lon_max)
# Order matters: smaller/more specific states first so they win over overlapping larger ones
BUNDESLAENDER = [
    ("Bremen",                   53.01, 53.23,  8.48,  8.99),
    ("Hamburg",                  53.39, 53.75,  9.73, 10.33),
    ("Berlin",                   52.33, 52.68, 13.08, 13.77),
    ("Saarland",                 49.11, 49.64,  6.35,  7.41),
    ("Schleswig-Holstein",       53.35, 55.06,  8.30, 11.35),
    ("Mecklenburg-Vorpommern",   53.10, 54.70, 10.59, 14.45),
    ("Brandenburg",              51.35, 53.56, 11.26, 14.77),
    ("Sachsen-Anhalt",           51.00, 53.05, 10.56, 13.19),
    ("Thueringen",               50.20, 51.65,  9.87, 12.66),
    ("Sachsen",                  50.17, 51.70, 11.87, 15.05),
    ("Hessen",                   49.39, 51.66,  7.77, 10.24),
    ("Rheinland-Pfalz",         48.96, 50.95,  6.11,  8.51),
    ("Baden-Wuerttemberg",       47.53, 49.79,  7.51, 10.50),
    ("Bayern",                   47.27, 50.57,  8.97, 13.84),
    ("Niedersachsen",            51.29, 53.90,  6.65, 11.60),
    ("NRW",                      50.32, 52.53,  5.86,  9.47),
]

def classify(lat, lon):
    for name, la_min, la_max, lo_min, lo_max in BUNDESLAENDER:
        if la_min <= lat <= la_max and lo_min <= lon <= lo_max:
            return name
    return None

def main():
    path = "C:/dev/geoforensic-karte/data_backup/geoforensic_map_data.json"
    print(f"Loading {path} ...")
    with open(path) as f:
        d = json.load(f)

    nm = d["nm"]
    all_idx = nm.index("_all_points")
    points = d["p"]

    # Build name->index map, add missing Bundesländer to nm
    name_to_idx = {n: i for i, n in enumerate(nm)}
    for bl_name, *_ in BUNDESLAENDER:
        if bl_name not in name_to_idx:
            name_to_idx[bl_name] = len(nm)
            nm.append(bl_name)

    reassigned = 0
    unmatched = 0
    for p in points:
        if p[3] != all_idx:
            continue
        bl = classify(p[0], p[1])
        if bl:
            p[3] = name_to_idx[bl]
            reassigned += 1
        else:
            unmatched += 1

    print(f"Reassigned: {reassigned}")
    print(f"Unmatched (kept as _all_points): {unmatched}")

    # Update R (region stats)
    from collections import defaultdict
    stats = defaultdict(lambda: {"n": 0, "sum_v": 0.0, "worst": 0.0})
    for p in points:
        name = nm[p[3]]
        s = stats[name]
        s["n"] += 1
        s["sum_v"] += p[2]
        if abs(p[2]) > abs(s["worst"]):
            s["worst"] = p[2]

    new_R = []
    for name in nm:
        s = stats.get(name)
        if s and s["n"] > 0:
            new_R.append({
                "name": name,
                "n": s["n"],
                "mean": round(s["sum_v"] / s["n"], 2),
                "worst": round(s["worst"], 1),
                "bb": None,
            })
        else:
            new_R.append({"name": name, "n": 0, "mean": 0, "worst": 0, "bb": None})

    d["nm"] = nm
    d["R"] = new_R

    # Remove _all_points from nm/R if it has 0 points left
    all_count = stats.get("_all_points", {}).get("n", 0)
    if all_count == 0:
        # Remap indices: remove _all_points
        old_all_idx = all_idx
        nm_new = [n for n in nm if n != "_all_points"]
        old_to_new = {}
        new_i = 0
        for old_i, n in enumerate(nm):
            if n == "_all_points":
                continue
            old_to_new[old_i] = new_i
            new_i += 1
        for p in points:
            p[3] = old_to_new[p[3]]
        d["nm"] = nm_new
        d["R"] = [r for r in new_R if r["name"] != "_all_points"]
        print("Removed _all_points (0 remaining)")

    out = path
    with open(out, "w") as f:
        json.dump(d, f, separators=(",", ":"))
    print(f"Written to {out} ({len(points)} points, {len(d['nm'])} regions)")

    # Also copy to deploy location
    import shutil
    deploy = "C:/dev/geoforensic-karte/geoforensic_map_data.json"
    shutil.copy2(out, deploy)
    print(f"Copied to {deploy}")

if __name__ == "__main__":
    main()

File: fix_regions.js
/**
 * Reassign all points in geoforensic_map_data.json to correct Bundesland
 * using point-in-polygon with official boundary GeoJSON.
 */
const fs = require('fs');

const d = JSON.parse(fs.readFileSync('geoforensic_map_data.json', 'utf8'));
const g = JSON.parse(fs.readFileSync('bundeslaender.geo.json', 'utf8'));

// ---- Name mapping: GeoJSON name → data nm name ----
const geoToData = {
  'Baden-Württemberg': 'Baden-Wuerttemberg',
  'Bayern': 'Bayern',
  'Berlin': 'Berlin',
  'Brandenburg': 'Brandenburg',
  'Bremen': 'Bremen',
  'Hamburg': 'Hamburg',
  'Hessen': 'Hessen',
  'Mecklenburg-Vorpommern': 'Mecklenburg-Vorpommern',
  'Niedersachsen': 'Niedersachsen',
  'Nordrhein-Westfalen': 'NRW',
  'Rheinland-Pfalz': 'Rheinland-Pfalz',
  'Saarland': 'Saarland',
  'Sachsen': 'Sachsen',
  'Sachsen-Anhalt': 'Sachsen-Anhalt',
  'Schleswig-Holstein': 'Schleswig-Holstein',
  'Thüringen': 'Thueringen',
};

// ---- Point-in-polygon (ray casting) ----
function pointInRing(lat, lon, ring) {
  // ring is [[lon,lat], [lon,lat], ...]
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const xi = ring[i][0], yi = ring[i][1];
    const xj = ring[j][0], yj = ring[j][1];
    if (((yi > lat) !== (yj > lat)) &&
        (lon < (xj - xi) * (lat - yi) / (yj - yi) + xi)) {
      inside = !inside;
    }
  }
  return inside;
}

function pointInPolygon(lat, lon, geometry) {
  if (geometry.type === 'Polygon') {
    // First ring is exterior, rest are holes
    const coords = geometry.coordinates;
    if (!pointInRing(lat, lon, coords[0])) return false;
    for (let h = 1; h < coords.length; h++) {
      if (pointInRing(lat, lon, coords[h])) return false; // inside a hole
    }
    return true;
  }
  if (geometry.type === 'MultiPolygon') {
    for (const poly of geometry.coordinates) {
      if (!pointInRing(lat, lon, poly[0])) continue;
      let inHole = false;
      for (let h = 1; h < poly.length; h++) {
        if (pointInRing(lat, lon, poly[h])) { inHole = true; break; }
      }
      if (!inHole) return true;
    }
    return false;
  }
  return false;
}

// ---- Build lookup: for each feature, compute bounding box for fast pre-filter ----
const states = g.features.map(f => {
  const dataName = geoToData[f.properties.name];
  if (!dataName) throw new Error('No mapping for: ' + f.properties.name);

  // Compute bbox
  let minLat = 999, maxLat = -999, minLon = 999, maxLon = -999;
  function scanCoords(coords) {
    if (typeof coords[0] === 'number') {
      if (coords[1] < minLat) minLat = coords[1];
      if (coords[1] > maxLat) maxLat = coords[1];
      if (coords[0] < minLon) minLon = coords[0];
      if (coords[0] > maxLon) maxLon = coords[0];
      return;
    }
    for (const c of coords) scanCoords(c);
  }
  scanCoords(f.geometry.coordinates);

  return {
    name: f.properties.name,
    dataName,
    dataIdx: d.nm.indexOf(dataName),
    geometry: f.geometry,
    bbox: { minLat, maxLat, minLon, maxLon }
  };
});

// Verify all data names are mapped
for (const s of states) {
  if (s.dataIdx === -1) throw new Error('Data name not found in nm: ' + s.dataName);
}

// ---- Reassign all points ----
console.log('Processing', d.p.length, 'points...');
let changed = 0, unmatched = 0;
const changeCounts = {}; // "from → to" counts

for (let i = 0; i < d.p.length; i++) {
  const pt = d.p[i];
  const lat = pt[0], lon = pt[1];
  const oldIdx = pt[3];

  // Find correct state
  let newIdx = -1;
  for (const s of states) {
    // Fast bbox check
    if (lat < s.bbox.minLat || lat > s.bbox.maxLat ||
        lon < s.bbox.minLon || lon > s.bbox.maxLon) continue;
    if (pointInPolygon(lat, lon, s.geometry)) {
      newIdx = s.dataIdx;
      break;
    }
  }

  if (newIdx === -1) {
    // Point outside Germany — find nearest state
    unmatched++;
    // Keep original assignment for now
    continue;
  }

  if (newIdx !== oldIdx) {
    const key = d.nm[oldIdx] + ' → ' + d.nm[newIdx];
    changeCounts[key] = (changeCounts[key] || 0) + 1;
    pt[3] = newIdx;
    changed++;
  }
}

console.log('\nChanged:', changed, '/', d.p.length, 'points');
console.log('Unmatched (outside polygons):', unmatched);
console.log('\nTransfers:');
Object.entries(changeCounts)
  .sort((a, b) => b[1] - a[1])
  .forEach(([k, v]) => console.log('  ' + k + ': ' + v));

// ---- Recalculate region stats (R array) ----
const regStats = {};
for (const name of d.nm) regStats[name] = { pts: [], minLat: 999, maxLat: -999, minLon: 999, maxLon: -999 };

for (const pt of d.p) {
  const name = d.nm[pt[3]];
  const s = regStats[name];
  s.pts.push(pt[2]); // velocity
  if (pt[0] < s.minLat) s.minLat = pt[0];
  if (pt[0] > s.maxLat) s.maxLat = pt[0];
  if (pt[1] < s.minLon) s.minLon = pt[1];
  if (pt[1] > s.maxLon) s.maxLon = pt[1];
}

d.R = d.nm.map(name => {
  const s = regStats[name];
  if (!s.pts.length) return { name, n: 0, mean: 0, worst: 0, bb: [0,0,0,0] };
  const mean = s.pts.reduce((a, b) => a + b, 0) / s.pts.length;
  const worst = s.pts.reduce((a, b) => Math.abs(b) > Math.abs(a) ? b : a, 0);
  return {
    name,
    n: s.pts.length,
    mean: Math.round(mean * 10) / 10,
    worst: Math.round(worst * 10) / 10,
    bb: [s.minLat, s.minLon, s.maxLat, s.maxLon]
  };
});

console.log('\nUpdated region stats:');
d.R.sort((a, b) => a.mean - b.mean);
d.R.forEach(r => console.log('  ' + r.name + ': n=' + r.n + ' mean=' + r.mean + ' worst=' + r.worst));

// ---- Write output ----
fs.writeFileSync('geoforensic_map_data.json', JSON.stringify(d));
console.log('\nSaved updated geoforensic_map_data.json');
